runtime/io/common/src/aws/smithy/kotlin/runtime/io/SdkByteReadChannel.kt

```diff
@@ -97,3 +100,34 @@ public suspend fun SdkByteReadChannel.readToBuffer(): SdkBuffer {
     readRemaining(buffer)
     return buffer
 }
+
+/**
+ * Read all bytes from this channel into [sink]. Returns the total number of bytes written.
+ */
+public suspend fun SdkByteReadChannel.readAll(sink: SdkSink): Long = withContext(SdkDispatchers.IO) {
+    val bufferedSink = if (sink is SdkBuffer) sink else sink.buffer()
+    var totalWritten = 0L
+    while (true) {
+        val rc = read(bufferedSink.buffer, DEFAULT_BYTE_CHANNEL_MAX_BUFFER_SIZE.toLong())
+        if (rc == -1L) break
+        totalWritten += rc
+        bufferedSink.emit()
+    }
+    bufferedSink.emit()
```

Comment:
**Question**: Remind me why we `emit` inside the loop and again after the loop? We didn't read anything between the last loop `emit` and exiting the loop, right?

Reply:
I don't recall of top of my head. I want to say there is some state you can get into where you have some data leftover from the last read but it's not obvious from looking at the code 🤔 . I can try taking it out and see what breaks. (It's not going to hurt anything though to leave it in)